xsdata/formats/dataclass/parsers.py

```diff
@@ -80,87 +82,97 @@ def parse(self, source: BytesIO, clazz: Type) -> Type:
 
-        return obj
-
-    def find_field(
-        self, queue: List[Dict], namespace: Optional[str], element: Element
-    ) -> Field:
-        """
-        Find the current field from the fields queue.
-
-        If the next_letter field is also a dataclass append its fields
-        map to the queue for the next_letter event
-        """
-        field = queue[-1][element.tag]
-        if field.is_dataclass:
-            class_fields = self.class_ns_fields(field.type, namespace)
-            queue.append(class_fields)
+        if len(queue) > 0 or len(stack) != 1:
+            raise AssertionError("Parsing failed with residual metadata")
 
-        return field
+        _, obj = stack.pop()
 
-    def build_object_from_field(
-        self, field: Field, namespace: Optional[str], element: Element
-    ) -> Type:
-        """Bind the current element to a dataclass or simply parse its text
-        value."""
-        if not field.is_dataclass:
-            return self.parse_value(field.type, element.text)
-
-        return self.build_object(field.type, namespace, element)
+        return obj
 
     def build_object(
-        self, clazz: Type, namespace: Optional[str], element: Element
-    ) -> Type:
-        """Create a new class instance by the current element attributes and
-        text."""
-        params = {}
-        for qname, field in self.class_ns_fields(clazz, namespace).items():
-            if field.is_text and element.text:
-                params[field.name] = self.parse_value(field.type, element.text)
-            elif qname in element.attrib:
-                params[field.name] = self.parse_value(field.type, element.attrib[qname])
+        self, item: QueueItem, element: Element, stack: List[Tuple[QName, Any]]
+    ) -> Tuple[QName, Any]:
+        """
+        Objectify current element by the item clazz type.
 
-        return clazz(**params)
+        If the clazz is a dataclass build the objects tree and parse attributes from
+        the element. Otherwise parse the elements text value
 
-    def end_element(
-        self, objects: List[Type], queue: List[Dict], element: Element
-    ) -> Type:
+        :returns Tuple: The qualified object name and the object
         """
-        Finalize and return the last item of the objects list.
-
-        Steps:
-           * Pop the last item of the objects
-           * If the object is a dataclass pop the fields queue which should be
-             the current object's fields map
-           * If the object is not the last in the list assign or append it
-             to the correct parent field
+        if self.is_dataclass(item.clazz):
+            children = self.fetch_class_children(item, stack)
+            attributes = self.parse_element_attributes(item, element)
+            obj = item.clazz(**children, **attributes)
+        else:
+            obj = self.parse_value(item.clazz, element.text)
+
+        return item.qname, obj
+
+    def fetch_class_children(
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/manual/tefra/xsdata/pullRequest?prid=4926385)